// cypress-html-validate/dist/commands.js
"use strict";
/// <reference types="cypress" />
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("./utils");
function logError(error, $el) {
    Cypress.log({
        $el,
        name: error.ruleId,
        consoleProps: () => error,
        message: [error.message],
    });
}
function findElements(messages, document) {
    return messages.map((message) => {
        return Object.assign(Object.assign({}, message), { element: message.selector ? document.querySelector(message.selector) : null });
    });
}
function matchSelectors(message, selector) {
    const target = message.element;
    return selector.some((elements) => {
        return elements.some((inner) => inner.contains(target));
    });
}
function filterMessages(messages, document, options) {
    const exclude = options.exclude.map((selector) => Array.from(document.querySelectorAll(selector)));
    const include = options.include.map((selector) => Array.from(document.querySelectorAll(selector)));
    return messages.filter((message) => {
        /* messages without an element cannot be filtered, assumed to be relevant */
        if (!message.element) {
            return true;
        }
        /* drop message if it is matched against excluded selector */
        if (exclude.length > 0 && matchSelectors(message, exclude)) {
            return false;
        }
        /* drop message if none of include selectors matches */
        if (include.length > 0) {
            return matchSelectors(message, include);
        }
        return true;
    });
}
function omitElement(message) {
    /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
    const { element } = message, rest = __rest(message, ["element"]);
    return rest;
}
function htmlvalidate() {
    Cypress.log({
        name: "html-validate",
        message: ["Validating document"],
    });
    cy.task("htmlvalidate:options", null, { log: false }).then((options) => {
        cy.document({ log: false })
            .then((document) => {
            const source = utils_1.getPageSource(document);
            return cy
                .task("htmlvalidate", source, { log: false })
                .then((report) => {
                if (report.valid) {
                    return cy.wrap([0, 0], { log: false });
                }
                else {
                    const messages = findElements(report.results[0].messages, document);
                    const filtered = filterMessages(messages, document, options);
                    cy.wrap(filtered, { log: false }).each((error) => {
                        const $el = error.element ? Cypress.$(error.element) : undefined;
                        logError(error, $el);
                    });
                    /* log violations to nodejs process */
                    if (filtered.length > 0) {
                        /* some properties in `HTMLElement` might not serialize
                         * properly so they are omitted from the messages before
                         * passing it to the nodejs process */
                        const stripped = filtered.map(omitElement);
                        cy.task("htmlvalidate:format", stripped);
                    }
                    return cy.wrap([messages.length, filtered.length], {
                        log: false,
                    });
                }
            });
        })
            .then(([messages, filtered]) => {
            const s = filtered !== 1 ? "s" : "";
            const excluded = messages - filtered;
            assert.equal(filtered, 0, `${filtered} error${s}, ${excluded} excluded`);
        });
    });
}
Cypress.Commands.add("htmlvalidate", htmlvalidate);
